File: hard/1096-brace-expansion-ii/solution.py
from typing import List, Set


class Solution:
    def braceExpansionII(self, expression: str) -> List[str]:
        """
        1096. Brace Expansion II
        Time: O(total output size * word length)
        Space: O(total output size)
        """
        n = len(expression)
        self.pos = 0

        def parse_union() -> Set[str]:
            words = parse_concat()
            while self.pos < n and expression[self.pos] == ",":
                self.pos += 1  # skip ','
                words |= parse_concat()
            return words

        def parse_concat() -> Set[str]:
            words = {""}
            while self.pos < n and expression[self.pos] not in ",}":
                part = parse_term()
                words = {a + b for a in words for b in part}
            return words

        def parse_term() -> Set[str]:
            if expression[self.pos] == "{":
                self.pos += 1  # skip '{'
                words = parse_union()
                self.pos += 1  # skip '}'
                return words
            ch = expression[self.pos]
            self.pos += 1
            return {ch}

        return sorted(parse_union())

File: hard/1096-brace-expansion-ii/solution.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

/**
 * 1096. Brace Expansion II
 * Time: O(total output size * word length)
 * Space: O(total output size)
 */
class Solution {
    private String expr;
    private int pos;

    public List<String> braceExpansionII(String expression) {
        expr = expression;
        pos = 0;
        return new ArrayList<>(new TreeSet<>(parseUnion()));
    }

    /** union := concat (',' concat)* */
    private Set<String> parseUnion() {
        Set<String> words = parseConcat();
        while (pos < expr.length() && expr.charAt(pos) == ',') {
            pos++;  // skip ','
            words.addAll(parseConcat());
        }
        return words;
    }

    /** concat := term+ */
    private Set<String> parseConcat() {
        Set<String> words = new HashSet<>();
        words.add("");
        while (pos < expr.length() && expr.charAt(pos) != ',' && expr.charAt(pos) != '}') {
            Set<String> part = parseTerm();
            Set<String> merged = new HashSet<>();
            for (String a : words) {
                for (String b : part) {
                    merged.add(a + b);
                }
            }
            words = merged;
        }
        return words;
    }

    /** term := letter | '{' union '}' */
    private Set<String> parseTerm() {
        if (expr.charAt(pos) == '{') {
            pos++;  // skip '{'
            Set<String> words = parseUnion();
            pos++;  // skip '}'
            return words;
        }
        Set<String> single = new HashSet<>();
        single.add(String.valueOf(expr.charAt(pos)));
        pos++;
        return single;
    }
}
